sdk/core/azure-core/src/http/websockets/websockets_impl.cpp

```diff
@@ -0,0 +1,884 @@
+    remoteOrigin += "://";
+    remoteOrigin += m_remoteUrl.GetHost();
+    openSocketRequest.SetHeader("Origin", remoteOrigin);
+
+    // Send the connect request to the WebSocket server.
+    auto response = openPipeline.Send(openSocketRequest, context);
+
+    // Ensure that the server thinks we're switching protocols. If it doesn't,
+    // fail immediately.
+    if (response->GetStatusCode() != Azure::Core::Http::HttpStatusCode::SwitchingProtocols)
+    {
+      throw Azure::Core::Http::TransportException("Unexpected handshake response");
+    }
+
+    // Prove that the server received this socket request.
+    auto& responseHeaders = response->GetHeaders();
+    if (!m_transport->HasBuiltInWebSocketSupport())
+    {
+      auto socketAccept(responseHeaders.find("Sec-WebSocket-Accept"));
+      if (socketAccept == responseHeaders.end())
+      {
+        throw Azure::Core::Http::TransportException("Missing Sec-WebSocket-Accept header");
+      }
+      // Verify that the WebSocket server received *this* open request.
+      else
+      {
+        VerifySocketAccept(encodedKey, socketAccept->second);
+      }
+      m_initialBodyStream = response->ExtractBodyStream();
+      m_pingThread.Start(m_transport);
+    }
+
+    // Remember the protocol that the client chose.
+    auto chosenProtocol = responseHeaders.find("Sec-WebSocket-Protocol");
+    if (chosenProtocol != responseHeaders.end())
+    {
+      m_chosenProtocol = chosenProtocol->second;
+    }
+
+    m_state = SocketState::Open;
+  }
+  bool WebSocketImplementation::HasBuiltInWebSocketSupport()
+  {
+    std::lock_guard<std::mutex> lock(m_stateMutex);
+    m_stateOwner = std::this_thread::get_id();
+    if (m_state != SocketState::Open)
+    {
+      throw std::runtime_error("Socket is not open.");
+    }
+    return m_transport->HasBuiltInWebSocketSupport();
+  }
+
+  std::string const& WebSocketImplementation::GetNegotiatedProtocol()
+  {
+    std::lock_guard<std::mutex> lock(m_stateMutex);
+    m_stateOwner = std::this_thread::get_id();
+    if (m_state != SocketState::Open)
+    {
+      throw std::runtime_error("Socket is not open.");
+    }
+    return m_chosenProtocol;
+  }
+
+  void WebSocketImplementation::AddHeader(std::string const& header, std::string const& headerValue)
+  {
+    std::lock_guard<std::mutex> lock(m_stateMutex);
+    m_stateOwner = std::this_thread::get_id();
+    if (m_state != SocketState::Closed && m_state != SocketState::Invalid)
+    {
+      throw std::runtime_error("AddHeader can only be called on closed sockets.");
+    }
+    m_headers.emplace(std::make_pair(header, headerValue));
+  }
+
+  void WebSocketImplementation::Close(
+      uint16_t closeStatus,
+      std::string const& closeReason,
+      Azure::Core::Context const& context)
+  {
+    std::unique_lock<std::mutex> lock(m_stateMutex);
+    m_stateOwner = std::this_thread::get_id();
+
+    // If we're closing an already closed socket, we're done.
+    if (m_state == SocketState::Closed)
```

Comment:
Does it make sense to error out if we are SocketState::Closing?  OR should that be considered the same as Closed and just return?

Reply:
Closing a closed socket cannot be an error, because sockets can be closed by the remote peer.

If closing a closed socket is an error, then a remote close (which is out of the application's control) will cause the app to crash due to circumstances beyond its control.

One other alternative is to maintain two states: A connection state and a socket state.  Then a remote teardown closes sets the connection state to closed even though the socket is still open. But I still need to have the "WebSOcket::Close is allowed on a closed connection" semantic.